pkg/kubelet/cm/deviceplugin/device_plugin_handler.go

```diff
@@ -0,0 +1,289 @@
+		// No change, no work.
+		return nil, nil
+	}
+	if needed < 0 {
+		// Pod is downsizing its resource requests. We need to make Allocate rpc call
+		// again to re-generate device runtime config. allocatedDevices and podDevices
+		// will be updated after Allocate rpc call succeeds. If the rpc call fails, we
+		// will keep the current cached state unchanged.
+		for devices.Len() > required {
+			devices.PopAny()
+		}
+		glog.V(3).Infof("Return downsized devices %v", devices.List())
+		return devices, nil
+	}
+	// Needs to allocate additional devices.
+	if h.allocatedDevices[resource] == nil {
+		h.allocatedDevices[resource] = sets.NewString()
+	}
+	// Gets Devices in use.
+	devicesInUse := h.allocatedDevices[resource]
+	// Gets a list of available devices.
+	available := h.allDevices[resource].Difference(devicesInUse)
+	if int(available.Len()) < needed {
+		return nil, fmt.Errorf("requested number of devices unavailable for %s. Requested: %d, Available: %d", resource, needed, available.Len())
+	}
+	allocated := available.UnsortedList()[:needed]
+	// Updates h.allocatedDevices with allocated devices to prevent them
+	// from being allocated to other pods/containers, given that we are
+	// not holding lock during the rpc call.
+	for _, device := range allocated {
+		h.allocatedDevices[resource].Insert(device)
+		devices.Insert(device)
+	}
+	return devices, nil
+}
+
+// Allocate attempts to allocate all of required extended resources for
+// the input container, issues an Allocate rpc request for each of such
+// resources, processes their AllocateResponses, and updates the cached
+// containerDevices on success.
+func (h *HandlerImpl) Allocate(pod *v1.Pod, container *v1.Container, activePods []*v1.Pod) error {
+	h.updateAllocatedDevices(activePods)
+	podUID := string(pod.UID)
+	contName := container.Name
+	for k, v := range container.Resources.Limits {
+		resource := string(k)
+		needed := int(v.Value())
+		glog.V(3).Infof("needs %d %s", needed, resource)
+		if _, registeredResource := h.allDevices[resource]; !registeredResource {
+			continue
+		}
+		allocDevices, err := h.devicesToAllocate(podUID, contName, resource, needed)
+		if err != nil {
+			return err
+		}
+		if allocDevices == nil || len(allocDevices) <= 0 {
+			continue
+		}
+		// devicePluginManager.Allocate involves RPC calls to device plugin, which
+		// could be heavy-weight. Therefore we want to perform this operation outside
+		// mutex lock. Note if Allcate call fails, we may leave container resources
+		// partially allocated for the failed container. We rely on updateAllocatedDevices()
+		// to garbage collect these resources later. Another side effect is that if
+		// we have X resource A and Y resource B in total, and two containers, container1
+		// and container2 both require X resource A and Y resource B. Both allocation
+		// requests may fail if we serve them in mixed order.
+		// TODO: may revisit this part later if we see inefficient resource allocation
```

Comment:
This can result in a race - Imagine a pod getting rejected by the kubelet because it temporarily stranded resources that it has previously allocated to a pod it had rejected because Allocation failed.

Reply:
May be run an update before every Allocation and ensure that the update is cheap?